pkg/cache/node.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2019 Cloudera, Inc.  All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"github.com/cloudera/yunikorn-core/pkg/api"
+	"github.com/cloudera/yunikorn-core/pkg/log"
+	"github.com/cloudera/yunikorn-k8shim/pkg/common"
+	"github.com/cloudera/yunikorn-k8shim/pkg/common/events"
+	"github.com/cloudera/yunikorn-k8shim/pkg/conf"
+	"github.com/cloudera/yunikorn-scheduler-interface/lib/go/si"
+	"github.com/looplab/fsm"
+	"go.uber.org/zap"
+	"sync"
+)
+
+// stores info about what scheduler cares about a node
+type SchedulerNode struct {
+	name      string
+	uid       string
+	capacity  *si.Resource
+	existingAllocations []*si.Allocation
+	schedulerApi api.SchedulerApi
+	fsm       *fsm.FSM
+	lock      *sync.RWMutex
+}
+
+func newSchedulerNode(nodeName string, nodeUid string,
+	nodeResource *si.Resource, schedulerApi api.SchedulerApi) *SchedulerNode {
+	schedulerNode := &SchedulerNode{
+		name: nodeName,
+		uid:  nodeUid,
+		capacity: nodeResource,
+		schedulerApi: schedulerApi,
+		lock: &sync.RWMutex{},
+	}
+	schedulerNode.initFSM()
+	return schedulerNode
+}
+
+func (n *SchedulerNode) initFSM() {
+	var states = events.States().Node
+	n.fsm = fsm.NewFSM(
+		states.New,
+		fsm.Events{
+			{Name: string(events.RecoverNode),
+				Src: []string{states.New},
+				Dst: states.Recovering,
+			},
+			{Name: string(events.NodeAccepted),
+				Src: []string{states.Recovering},
+				Dst: states.Healthy,
+			},
+			{Name: string(events.NodeRejected),
+				Src: []string{states.New, states.Recovering},
+				Dst: states.Rejected,
+			},
+		},
+		fsm.Callbacks{
+			string(states.Recovering): n.handleNodeRecovery,
+		})
+}
+
+func (n *SchedulerNode) addExistingAllocation(allocation *si.Allocation) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.existingAllocations = append(n.existingAllocations, allocation)
+}
+
+func (n *SchedulerNode) getNodeState() string {
+	n.lock.RLock()
```

Comment:
According to the doc fsm has its own internal locking and we do not need to guard here.

Reply:
Yes. Make sense. I just removed the lock for node, and task. app and scheduler are fine.